### PR TITLE
test: migrate to vitest + add fork/reorg coverage suite

### PR DESCRIPTION
**Pull request by mo4islona**
## Summary

Migrates every package's test runner from mocha to vitest and lands a comprehensive fork/reorg coverage suite across the processor stack.

- **vitest migration** for all packages, with short vs LFS suites split where applicable.
- **`@subsquid/util-internal-testing`** — new package with in-process `MockPortal`, `MockChain` (deterministic block / fork builders), and `MockDataSource` for HotProcessor unit tests.
- **HotProcessor unit tests** (`util-internal-ingest-tools/src/hot.test.ts`) — linear progress, shallow/deep/cascading forks, divergent input, deep `top[]` init, `finalize()` edge cases (past-tip clamp, genesis `||`-trap, wrong-hash at known height), finality-only update tripwire.
- **batch-processor** — `findRollbackIndex` unit tests (disjoint, empty chain, below-chain `previousBlocks`); end-to-end fork scenarios via `ProgrammableSource`.
- **util-internal-processor-tools** — Runner integration tests for the legacy hot path (HotProcessor + transactHot2 driven from in-process data sources).
- **typeorm-store** — 11 new test files (`fork-deep`, `fork-cascade`, `hot-edge-cases`, `mixed-boundary`, `data-type-rollback`, `transact-parity`, `templates-hot`, `concurrency`, `crash-recovery`, `long-top`, `cockroach-smoke`). Run against real Postgres + CockroachDB via docker-compose.
- **Chain-specific RPC** unit tests — EVM `ensureContinuity`, Bitcoin `ensureContinuity` (incl. `ZERO_HASH` genesis), Solana `ChainFixer` (`acceptBatch` / `fillGap` / `removeOverlaps` / `limitUpperBoundary`).
- **`it.fails` tripwires** for documented live bugs flip green automatically once the underlying issue is fixed.

## Test plan

- [ ] \`pnpm -F @subsquid/util-internal-ingest-tools test\` — HotProcessor unit suite passes (no DB / no network).
- [ ] \`pnpm -F @subsquid/util-internal-testing test\` — mock infrastructure smoke.
- [ ] \`pnpm -F @subsquid/batch-processor test\` — findRollbackIndex unit + Processor scenarios.
- [ ] \`pnpm -F @subsquid/util-internal-processor-tools test\` — Runner integration suite.
- [ ] \`pnpm -F @subsquid/evm-rpc test\` — EVM ensureContinuity unit + existing suites.
- [ ] \`pnpm -F @subsquid/bitcoin-rpc test\` — Bitcoin ensureContinuity unit.
- [ ] \`pnpm -F @subsquid/solana-rpc test\` — Solana ChainFixer unit.
- [ ] \`cd typeorm/typeorm-store && make up && make test\` — full DB-backed suite on Postgres and CockroachDB.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: docs/testing/processor-test-strategy.md [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+| HotProcessor | Links new headers backward (by parent-hash). If they diverge from the local chain — replaces the divergent suffix. Emits `HotUpdate.baseHead` = the common ancestor. |
+| Runner / Processor | Consumes `HotUpdate` or `ForkException`; in the new `Processor`, `handleFork` truncates `unfinalizedHeads` to the common ancestor. |
+| Database (typeorm-store) | In `transactHot2` finds `baseHead` in `[state, ...state.top]`, deletes everything above it and replays `rollbackBlock` across the change-log (reverse playback). |
+
+---
+
+## 2. Invariants (what we verify)
+
+1. **`HotDatabaseState.top[]` monotonicity** — heights are strictly increasing (`top[i].height > top[i-1].height`), and each block's parent-hash links to its predecessor. **Height gaps are legal** (Solana skipped slots, other chains with sparse slots). The `assertChainContinuity` guard intentionally uses `>`, not `===`. What must never happen: duplicate heights, reversed heights, or a missing hash-link.
```
[the remark this answers is not in the document]
Comment (mo4islona) [reply docs/testing/processor-test-strategy.md]: Accepted — fixed in 86f07d24. Verified against code: `hot.ts:35` uses `==` (strictly consecutive), `database.ts:372` uses `>` (monotonic only). Invariant #1 now splits by layer and notes that the gap-tolerant boundary is strictly below HotProcessor.

File: docs/testing/processor-test-strategy.md [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+2. **`baseHead` ∈ current chain** — in every `HotUpdate`, `baseHead` is either `state` itself or some element of `state.top[]`. A violation is a HotProcessor bug.
+3. **Change-log ordering** — rollback reads `ORDER BY index DESC`; operation mapping: `insert→DELETE`, `update→UPDATE prev`, `delete→INSERT prev`. After a full rollback the user-table state equals the state before the block's first write.
+4. **Status `nonce`** — monotonic. A concurrent writer gets 0 updated rows and the error `'status table was updated by foreign process'` (`database.ts:344`).
```
[the remark this answers is not in the document]
Comment (mo4islona) [reply docs/testing/processor-test-strategy.md]: Accepted — fixed in 86f07d24. Invariant #4 now quotes the full `RACE_MSG` string so tests can grep/assert on it literally.

File: docs/testing/processor-test-strategy.md [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+interface HotDatabase<S> {
+    supportsHotBlocks: true
+    connect(): Promise<HotDatabaseState>           // {height, hash, top: HashAndHeight[]}
+    transact(info: FinalTxInfo, cb): Promise<void>
+    transactHot(info: HotTxInfo, cb): Promise<void>
+    transactHot2?(info: HotTxInfo, cb): Promise<void>
+}
+```
+
+Implementation — `typeorm-store` (`typeorm/typeorm-store/src/database.ts`, 341 lines):
```
[the remark this answers is not in the document]
Comment (mo4islona) [reply docs/testing/processor-test-strategy.md]: Accepted — fixed in 86f07d24. "341 lines" already drifted to 376; dropped it. Kept the code-anchored line-range references (e.g. "lines 194–258") since they're useful for navigation.

File: docs/testing/processor-test-strategy.md [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+
+### 1.3 Where forks are detected and resolved
+
+| Layer | Role |
+|---|---|
+| RPC data source | Compares `block.parentHash` with the expected value; on mismatch throws `ForkException` carrying `expectedParentHash` + `previousBlocks` (blocks already emitted downstream before the fork). |
+| HotProcessor | Links new headers backward (by parent-hash). If they diverge from the local chain — replaces the divergent suffix. Emits `HotUpdate.baseHead` = the common ancestor. |
+| Runner / Processor | Consumes `HotUpdate` or `ForkException`; in the new `Processor`, `handleFork` truncates `unfinalizedHeads` to the common ancestor. |
+| Database (typeorm-store) | In `transactHot2` finds `baseHead` in `[state, ...state.top]`, deletes everything above it and replays `rollbackBlock` across the change-log (reverse playback). |
+
```
[the remark this answers is not in the document]
Comment (mo4islona) [reply docs/testing/processor-test-strategy.md]: Dismissed — the tables use single leading `|` (verified with `grep -n '^||'` returning empty). The visual ambiguity likely comes from the em-dash `—` columns adjacent to `|`; nothing to fix.

File: docs/testing/processor-test-strategy.md [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+---
+
+## 3. Case matrix (scenarios × layers)
+
+Legend: ✅ = covered, 🆕 = to add, — = not applicable.
+
+| # | Scenario | HotProcessor (unit) | typeorm-store (integ) | Runner (integ, mock portal) | batch-processor (integ, mock portal) | RPC DS (unit) |
+|---|---|:-:|:-:|:-:|:-:|:-:|
+| 1 | Linear progress, all finalized | — | ✅ | 🆕 | 🆕 | — |
```
[the remark this answers is not in the document]
Comment (mo4islona) [reply docs/testing/processor-test-strategy.md]: Dismissed — duplicate of the first table comment. Tables use single leading `|`; the rendering issue is cosmetic from adjacent em-dashes, not malformed markdown.

File: rush.json [hunk 6]
```diff
@@ -976,6 +976,12 @@
       "shouldPublish": true,
       "versionPolicyName": "npm"
     },
+    {
+      "packageName": "@subsquid/util-internal-testing",
+      "projectFolder": "util/util-internal-testing",
+      "shouldPublish": true,
```
Comment (belopash) [inline rush.json]: false?
